### PR TITLE
Add support for granite models

### PR DESCRIPTION
**Pull request by mariofusco**
I'm trying to add support for `granite` models following @tjake's suggestions, but this doesn't seem to work as expected. I'm trying to use [this model](https://huggingface.co/ibm-granite/granite-3.0-2b-instruct), but when I try to chat with it, I only get garbage.

Screenshot from 2024-10-25 17-09-41

I also tried to quantize the model, but the outcome doesn't change. I believe that there is some subtle difference between the [llama](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py) and [granite](https://github.com/huggingface/transformers/blob/main/src/transformers/models/granite/modeling_granite.py) transformers implementation, but so far giving a look at the linked corresponding python transformers I cannot understand where it is.

Any suggestion on how to improve this is welcome.

### CODE REVIEW

**Comment by tjake** [comment]
The issue seems to be the tokenizer, I will take a look

**Comment by tjake** [comment]
Hi @mariofusco,

Well this was trickier than initially appeared.  

The model has a few differences from llama:

1. Includes multipliers for embeddings, attention, residuals and logits
2. It does not have a Beginning of sentence token which is very odd
3. The tokenizer is GPT2Tokenizer not LlamaTokenizer
4. Finally the model is missing a tokenizer.json (I've left a [comment on the model](https://huggingface.co/ibm-granite/granite-3.0-2b-instruct/discussions/5) so they add it in the meantime use the one from the [8b model](https://huggingface.co/ibm-granite/granite-3.0-8b-instruct/blob/main/tokenizer.json))

Now the model works for me

**Comment by tjake** [comment]
Screenshot from 2024-10-27 01-41-08

**Comment by mariofusco** [comment]
> Now the model works for me

Amazing job, I will take a look immediately :+1:

**Comment by mariofusco** [comment]
@tjake Great work, much better :) I just have a few notes.

First trying to use the [granite model](https://huggingface.co/ibm-granite/granite-3.0-2b-instruct) that I linked also in my previous comment, I got a response like the following

image

... and it continue relentlessly printing more and more capitals till I killed the server. Do you think that this is expected? Something related to how the model has been designed? 

Second I also tried the quantized version of that model, but that one doesn't seem to work at all, never returning any result. To try to figure out what was going on there I also tried the same model with my [toy project](https://github.com/mariofusco/quarkus-drools-llm) and it caused the following exception:

```
java.lang.UnsupportedOperationException
        at com.github.tjake.jlama.tensor.operations.PanamaTensorOperations.scale(PanamaTensorOperations.java:2368)
        at com.github.tjake.jlama.tensor.operations.NativeTensorOperations.scale(NativeTensorOperations.java:439)
        at com.github.tjake.jlama.model.granite.GraniteModel.lambda$loadInputWeights$1(GraniteModel.java:121)
        at com.github.tjake.jlama.model.functions.EmbedInput.batchInputsToEmbeddings(EmbedInput.java:33)
        at com.github.tjake.jlama.model.AbstractModel.batchForward(AbstractModel.java:280)
        at com.github.tjake.jlama.model.AbstractModel.batchForward(AbstractModel.java:271)
        at com.github.tjake.jlama.model.AbstractModel.generate(AbstractModel.java:542)
        at io.quarkiverse.langchain4j.jlama.JlamaChatModel.generate(JlamaChatModel.java:98)
        at io.quarkiverse.langchain4j.jlama.JlamaChatModel.generate(JlamaChatModel.java:84)
        at dev.langchain4j.model.chat.ChatLanguageModel_XNMsOaekknG7BdNZ5YSUkjh1SqE_Synthetic_ClientProxy.generate(Unknown Source)
        at io.quarkiverse.langchain4j.runtime.aiservice.AiServiceMethodImplementationSupport.doImplement(AiServiceMethodImplementationSupport.java:269)
        at io.quarkiverse.langchain4j.runtime.aiservice.AiServiceMethodImplementationSupport.implement(AiServiceMethodImplementationSupport.java:125)
        at org.hybridai.refund.aiservices.CustomerChatService$$QuarkusImpl.chat(Unknown Source)
```

Note however that I performed that quantization before pulling your fixes. Is the quantization process affected by them, so maybe I should try to rerun it now?

That said I believe that this pull request is already a huge step forward, so I suggest to merge it and eventually add other minor adjustments in a follow up if necessary.

**Comment by mariofusco** [comment]
Sorry @tjake, ignore my first remark. That behavior is of course determined by the [tokenizer_config.json](https://huggingface.co/ibm-granite/granite-3.0-2b-instruct/blob/main/tokenizer_config.json) file for how it has been configured in that model. I'm learning how this stuff works only now, please be patient.

The second point, regarding the quantized model is still valid instead. I will try to rerun that quantization and see what happens.

**Comment by mariofusco** [comment]
> The second point, regarding the quantized model is still valid instead. I will try to rerun that quantization and see what happens.

Nope, I tried to rerun the quantization, but got exactly the same exception as before.

**Comment by tjake** [comment]
Hi Mario,

Did you use the tokenizer.json from the 8b model I linked? I had the same issue as you had when I used the 2b base tokenizer as its had some inconsistent values.

In anycase I quantized and uploaded the model to my huggingface please try that out `tjake/granite-3.0-2b-instruct-JQ4`

**Comment by tjake** [comment]
Here is the command I used to quantize:

`./run-cli.sh quantize --skip-layer=embed --skip-layer=norm ibm-granite/granite-3.0-2b-instruct`  I skip the embedding layer to avoid the error you hit (Q4 scale command isn't supported).

**Comment by mariofusco** [comment]
With that `skip-layer` flag it also works for me. Please feel free to merge this pull request if it's ok for you.

**Comment by tjake** [comment]
Thanks for doing this Mario!
